docs/specs/pipelines-io.md

```diff
@@ -0,0 +1,196 @@
+## Efficient IO
+
+As part of the high performance work, we need new abstractions for doing efficient reading and writing to and from the network. The pattern we have settled on
+so far in pipelines is a model where buffers are pushed by a producer (`IPipelineWriter`) to a consumer (`IPipelineReader`).
+
+```C#
+public interface IPipelineReader
+{
+    ReadableBufferAwaitable ReadAsync();
```

Comment:
I specifically avoided that and added CancelPendingRead because of having to register the callback per `ReadAsync` call. See https://github.com/dotnet/corefxlab/issues/1017 for more details.